File: algorithms/cpp/70.爬楼梯.cpp
#include <unordered_map>
using namespace std;
/*
 * @lc app=leetcode.cn id=70 lang=cpp
 *
 * [70] 爬楼梯
 */

/**
 * 递归并剪枝(缓存减少重复计算)
 */
class Solution2 {
 private:
  unordered_map<int, int> cache;

 public:
  int climbStairs(int n) {
    if (n < 3) return n;
    if (!cache.count(n)) {
      cache[n] = climbStairs(n - 1) + climbStairs(n - 2);
    }
    return cache.at(n);
  }
};

// @lc code=start
/**
 * 转化为 Fibonacci 问题
 */
class Solution {
 public:
  int climbStairs(int n) {
    int cur = 2, pre = 1;
    if (n <= 3) return n;
    for (int i = 2; i < n; i++) {
      int tmp = cur;
      cur += pre;
      pre = tmp;
    }
    return cur;
  }
};
// @lc code=end

File: algorithms/rs/70.爬楼梯.rs
/*
 * @lc app=leetcode.cn id=70 lang=rust
 *
 * [70] 爬楼梯
 */

// @lc code=start
impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut prev = 1;
        let mut cur = 2;
        if n < 3 {
            return n;
        }
        for i in 3..=n {
            let temp = cur;
            cur = cur + prev;
            prev = temp;
        }
        cur
    }
}
// @lc code=end

